### ISSUE TITLE
Files with "#" in name fail

### ISSUE THREAD

**Issue by nbbn** [MeanEYE/Sunflower#195]
Hello,
sunflower is great!
There is issue with handling files with "#" in name.
Only part of path before "#" is send to program, when file is double-clicked.
For example, when I try to open file "test#test" file "test" is opened.

**Comment by MeanEYE** [comment]
Thanks for kind words and for reporting this issue. Can you please provide version of Sunflower you are using. Just to make sure we are both on the same page. I remember similar issue to this one being reported recently so I'd like to make sure that issue is properly described.

**Comment by nbbn** [comment]
I use Sunflower 0.3(61) from deb package.
I updated the first post.

**Comment by ArseniyK** [comment]
I suppose it same problem https://github.com/MeanEYE/Sunflower/issues/66

**Comment by nbbn** [comment]
Next problems,
1. right click on file -> Open with -> another application -> in new window select gedit.
Then you get opened file: filename" (yes, quote at the end)

2. open text file test%test by double click in gedit. Opened file will be Untitled Document 1 with strange loading sign. F4 works fine.

**Comment by MeanEYE** [comment]
Thanks for testing.

**Comment by nbbn** [comment]
Next quirk:
I have file 36%160.pdf.
When I try to move using: ctrl+x, tab ctrl+v I receive message: Application is unable to handle specified data. Check if source items still exist.
When I use F6, it works fine.

**Comment by MeanEYE** [comment]
Oh, this sounds like a new issue as `Ctrl+X` puts the file path to clipboard. You should test by pressing `Ctrl+X` in Nautilus or something like that and then pressing `Ctrl+V` in Sunflower. Or the other way around.

**Comment by nbbn** [comment]
It works great between Nautilus, xfe and Sunflower. In both directions.
Maybe it is a new issue. Please make new issue, if you wish.

**Comment by MeanEYE** [comment]
It is a second issue.

**Comment by nervgh** [comment]
Same problem here.
In my case I open a video file via the Tothem player.
When I try to open a file with `#` in name, then the Sunflower opens another file in current directory.
I made a [video](https://drive.google.com/file/d/0B2CP8LcWJSNSVnhMZERtU3dGdE0) of this behavior.

**Comment by MeanEYE** [comment]
Thanks @nervgh.

**Comment by cmallwitz** [comment]
I looked into file names with a " in it: in order to execute e.g. the default editor, we take 'gedit %U' and replace %U with a double-quoted file name e.g. gedit "foo" But we don't escape a double-quote contained in the file name itself e.g. gedit "foo"bar" and then the launch process fails. gedit "foo\\bar" works as expected. This should be a simple change to associations.__format_command_string(). Should I just submit a fix?

**Comment by MeanEYE** [comment]
Yes @cmallwitz. That would be great.

**Comment by cmallwitz** [comment]
PR #205 submitted.
Christian

**Comment by cmallwitz** [comment]
Some further information for files containing # or % opened using (right mouse click) context menu - open: this hits associations.py open_file() line 212

For example file name /home/user/foo#bar.txt this is being turned into file:///home/user/foo#bar.txt - note: '#' and '%' have special meaning in a URI and therefore the file name is misinterpreted when gedit tries to use Gio to open the file (based on URI).

URL encoding special characters as in file:///home/user/foo%23bar.txt fixes the issue. Same with % being replaced with %25 first.

A more generic fix would be wrapping selections/paths using urllib.pathname2url() before prefixing it with file://

Any other suggestions? If not I can submit a fix.

**Comment by MeanEYE** [comment]
@cmallwitz sorry for delayed response. Basically paths should be URL encoded only if they are in form of URI complete with `file://`. We shouldn't blindly apply URL encode to all paths, instead we should see if while being executed application requires URI or path. As for using `Ctrl+C/V` (clipboard operations) as well with drag and drop we should universally use URIs.

**Comment by cmallwitz** [comment]
I completely agree with only encoding paths if the are URI complete with file:// - My fix is below and applicable in this situation:

diff --git a/application/associations.py b/application/associations.py
index 75857b2..03a434b 100644
--- a/application/associations.py
+++ b/application/associations.py
@@ -209,7 +209,7 @@ class AssociationManager:
 
 			if application is not None:
 				if application.supports_uris():
-					selection = map(lambda path: 'file://{0}'.format(path) if not path.startswith('file://') else path, selection)
+					selection = map(lambda path: 'file://{0}'.format(urllib.pathname2url(path)) if not path.startswith('file://') else path, selection)
 					application.launch_uris(selection)
 				else:
 					application.launch([Gio.File.new_for_path(path) for path in selection])

**Comment by cmallwitz** [comment]
Re-posting the diff part:

`
diff --git a/application/associations.py b/application/associations.py
index 75857b2..03a434b 100644
--- a/application/associations.py
+++ b/application/associations.py
@@ -209,7 +209,7 @@ class AssociationManager:
 
 			if application is not None:
 				if application.supports_uris():
-					selection = map(lambda path: 'file://{0}'.format(path) if not path.startswith('file://') else path, selection)
+					selection = map(lambda path: 'file://{0}'.format(urllib.pathname2url(path)) if not path.startswith('file://') else path, selection)
 					application.launch_uris(selection)
 				else:
 					application.launch([Gio.File.new_for_path(path) for path in selection])
`

**Comment by cmallwitz** [comment]
Re-posting the diff part:
`
diff --git a/application/associations.py b/application/associations.py
index 75857b2..03a434b 100644
--- a/application/associations.py
+++ b/application/associations.py
@@ -209,7 +209,7 @@ class AssociationManager:
 
 			if application is not None:
 				if application.supports_uris():
-					selection = map(lambda path: 'file://{0}'.format(path) if not path.startswith('file://') else path, selection)
+					selection = map(lambda path: 'file://{0}'.format(urllib.pathname2url(path)) if not path.startswith('file://') else path, selection)
 					application.launch_uris(selection)
 				else:
 					application.launch([Gio.File.new_for_path(path) for path in selection])
`

**Comment by cmallwitz** [comment]
This isn't pretty - I'm only suggesting changing one line - from:

selection = map(lambda path: 'file://{0}'.format(path) if not path.startswith('file://') else path, selection)

to:

selection = map(lambda path: 'file://{0}'.format(urllib.pathname2url(path)) if not path.startswith('file://') else path, selection)

**Comment by MeanEYE** [comment]
I think we should switch from manually formatting strings to using `urllib` and possible check if GTK application class has some sort of format parameters and get command line. Perhaps even execute with params.

**Comment by cmallwitz** [comment]
Submitted PR #216 to fix issue - fix is using urllib for formatting and only if required.

**Comment by cmallwitz** [comment]
This can be closed now.

**Comment by joshas** [comment]
@MeanEYE This issue can be closed, as it is fixed in develop branch.
